Rollback on failed unmount operations

kursatkobya opened the pull request:
CreateDir: Add a new class to track to hold directory creation backlog.
FileToolkitWithUndo: Use smart pointer instead of raw pointer.
Rollback on failed unmount operations
unittests for CreateDir class

Note : one tests marked to be skipped. It is necessary to mark it as "DISABLED" due to current CI conditions.

kursatkobya committed:
CreateDir: Add a new class to track to hold directory creation backlog.

Failing mounts in SoftwareContainer may cause unused directories. To
prevent this a backlog mechanism added to createDirectory function. With
this feauture rollback is possible on failure of createDirectory.

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

kursatkobya committed:
FileToolkitWithUndo: Use smart pointer instead of raw pointer.

member vector m_cleanupHandlers now holds unique pointers of
CleanUpHandler object.

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

kursatkobya committed:
FileToolkitWithUndo: Rollback on failed unmount operations

Delete and unmount unused/unnecessary directories on failure of BindMount.

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

kursatkobya committed:
Unittest: Add unittest for CreateDir class

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

tobsan commented on common/filetoolkitwithundo.cpp:
```diff
@@ -180,15 +123,33 @@ bool FileToolkitWithUndo::overlayMount(const std::string &lower,
                        lower << ", upperdir=" << upper << ", workdir=" << work;
         return false;
     }
+    m_create.clear();
```
I don't really like this solution at all. Wouldn't it make more sense if one could have a local FileToolKitWithUndo instance that also has a cleanup handler API?

```
FileToolkKitWithUndo ft;
if (!ft.createDirectory(lower) || ...) {
   // Something went wrong
} else 
    m_cleanupHandlers.emplace_back(ft);
}
```

tobsan replied:
In the went wrong step, the local ft would go out of scope and would automatically cleanup as part of its dtor.

kursatkobya replied:
You mean in overlayMount method of FileToolkitWithUndo class should create a inception FileToolkKitWithUndo class. I don't believe that would be nicer way. It does not make more sense to me, no.

tobsan commented on common/filetoolkitwithundo.h:
```diff
@@ -127,14 +109,9 @@ class FileToolkitWithUndo
-    bool createParentDirectory(const std::string &path);
+    std::vector<std::unique_ptr<CleanUpHandler>> m_cleanupHandlers;
+
+    CreateDir m_create;
```
Separating the directory creation code like this feels unnatural, and causes code duplication later on (either clean, or add the stuff done with m_create to a cleanuphandler anyway)

tobsan commented on common/unit-test/createdir_unittest.cpp:
```diff
@@ -0,0 +1,66 @@
+
+    CreateDir cd;
+
+};
```
It would make more sense to test parts of the FileToolkitWithUndo instead, in line with my other comments.

tobsan commented on common/unit-test/createdir_unittest.cpp:
```diff
@@ -0,0 +1,66 @@
+{
+    EXPECT_THROW(cd.tempDir("tmp/sc-unappropriate/template"), SoftwareContainerError);
+    ASSERT_TRUE(cd.rollBack());
+}
```
These tests are good though, this is the type of stuff we want to test with the cleanuphandlers.

tobsan replied:
Comments on all these test cases would be good.

tobsan replied:
None of these tests actually test that `rollBack()` removes the files, or that they do not already exist before running the tests.

Also, this is the only place where the `rollBack()` function is used as a public member. The tests could be rewritten so that `rollBack()` doesn't have to be public. Not really important though...

breakreturn replied:
@tobsan Isn't rollback() one of the main features? I'm probably misreading your comment...

tobsan replied:
@breakreturn yes, but rollback is done automatically upon destruction of the object.

breakreturn replied:
@tobsan Yeah, but I thought we needed that for some reason? Why would it otherwise exist at all on the public interface?

tobsan replied:
We need the functionality, but it can be triggered by just letting the object die, which will happen around the same time anyway in most cases. But it doesn't hurt to keep it.

kursatkobya replied:
We had one test examines `rollBack`. But as you know it is disabled for now due to root user issue on unit tests. Yet `rollBack` function can be removed as you mentioned.

breakreturn replied:
I think we should have as few methods as possible visible to the user and as crisp and clear interface possible as well. More methods means more ways to do the same thing, means more tests, means more ways to fail etc. I'd just like to see one clear explanation of how to use this class, if it's "might be needed but often happens at the same time", I think there is a risk it's not thought through.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -110,11 +110,16 @@ bool Container::initialize()
         std::string gatewayDir = gatewaysDir();
-        if (!createDirectory(gatewayDir)) {
+        if (!m_create.createDirectory(gatewayDir)) {
+            m_create.rollBack();
```
This change makes little sense as any parent directories would have been cleaned anyway when the container dtor is called (and this is kinda fatal error).

tobsan replied:
Also this has nothing to do with failing mounts.

kursatkobya replied:
ok it will be removed.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -110,11 +110,16 @@ bool Container::initialize()
+        m_create.clear();
+        if (!pathInList(gatewayDir)) {
+            m_cleanupHandlers.emplace_back(new DirectoryCleanUpHandler(gatewayDir));
+        }
```
Here is another example of how this change creates double checks instead of using the existing FileToolkitWithUndo.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -620,14 +627,40 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
+        rollback.clean();
+        // clean createdDirectories as well
+        m_create.rollBack();
+        return false;
```
This is going in the right direction though, but it could be done nicer.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -620,14 +627,40 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
+    return true;
```
This feels largely like we're copying parts of the code from the FileToolKitWithUndo and making it more specific. Do we have to?

tobsan commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -440,11 +440,16 @@ void SoftwareContainer::checkWorkspace()
+        m_create.clear();
+        if (!pathInList(rootDir)) {
+            m_cleanupHandlers.emplace_back(new DirectoryCleanUpHandler(rootDir));
+        }
```
This part of the code has nothing to do with failed mounts.

tobsan commented on common/createdir.h:
```diff
@@ -0,0 +1,94 @@
+     * @brief m_createDirectoryCleanupHandlers holds cleanup handler for each created directory.
+     * The main purpose of this is to rollback in any errors.
+     */
+    std::vector<std::unique_ptr<DirectoryCleanUpHandler>> m_rollbackCleaners;
```
Is there a reason for using pointers here and not objects or references directly?

tobsan commented on common/createdir.h:
```diff
@@ -0,0 +1,94 @@
+     * The main purpose of this is to rollback in any errors while bind-mounting. On success
+     * m_tempFileCleaners will be moved to cleanup handlers of FileToolkitWithUndo
+     */
+    std::vector<std::unique_ptr<DirectoryCleanUpHandler>> m_tempFileCleaners;
```
Should this member really be public?

tobsan reviewed (changes requested):
I find the CreateDir class completely superfluous. 

Furthermore, this PR does not solve the issues that might arise from a failing bind mount inside the container. What if the mount-move fails? What if a re-mount to read-only fails? These cases does not seem to be addressed here.

kursatkobya committed:
FileToolkitWithUndo: Change createDir and cleanup mechanisms.

CreateDir member changed to a vector of CreateDir pointers to be able to
create more robust and consise solution for mount rollbacks.

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

tobsan commented on common/createdir.cpp:
```diff
@@ -0,0 +1,116 @@
+    char *dir = const_cast<char*>(templ.c_str());
+    dir = mkdtemp(dir);
+    if (dir == nullptr) {
+        std::string message = "Failed to create buffered Directory: " + std::string(strerror(errno));
```
This isn't necessarily a buffered directory, it's just a tempdir.

tobsan commented on common/filetoolkitwithundo.h:
```diff
@@ -127,14 +109,8 @@ class FileToolkitWithUndo
-     */
-    bool createParentDirectory(const std::string &path);
+    std::vector<std::unique_ptr<CleanUpHandler>> m_cleanupHandlers;
+    std::vector<std::unique_ptr<CreateDir>> m_createDirList;
```
It would be good with some comments about why these are separated.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -622,12 +625,23 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
                 log_error() << "Could not create file " << tempPath;
                 return false;
             }
-            m_cleanupHandlers.push_back(new FileCleanUpHandler(tempPath));
```
Why is this line removed? If I read this correctly, this was the cleanup handler for a file that was touched in case we didn't mount to a dir but to a file. So why not cleanup that file just as we clean up the directories? Especially in the case where things go bad.

kursatkobya replied:
Here this is moved to end of the function. See the snippet

```
if (!pathIsDirectory) {
    m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
 }
```

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -622,12 +625,23 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
+    if (!bindMountCore(pathInHost, pathInContainer, tempPath, readOnly)) {
+        // bindMountInContainer operation is not successful. Clean mounted tempPath.
+        MountCleanUpHandler rollback{tempPath};
+        rollback.clean();
```
So we create a `MountCleanUpHandler` with `tempPath` and then instantly clear it of that path?

kursatkobya replied:
Yes, this is a line just for cleaning mount. Instead we can umount2 here, but i think it would be quite ugly.

tobsan replied:
Ah yes, I read it as a CreateDir object. My bad.

tobsan commented on common/createdir.cpp:
```diff
@@ -0,0 +1,116 @@
+void CreateDir::clear()
+{
+    m_rollbackCleaners.clear();
+}
```
This call is only used once in the code (and I'm not sure it should). Consider removing it.

tobsan reviewed (changes requested):
Looks better, see further comments.

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,90 @@
+     * @brief rollBack calls clean method for all members of m_createDirectoryCleanupHandlers.
+     *
+     * This function is created as a helper to callers of createDirectory functions. It is strongly
+     * recommended that this functions should be called if createDirectory function returns false.
```
I think comments like this also should be summarized, in the context of how to use this class, as a class comment above the class declaration.

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,90 @@
+     * @brief createDirectory Create a directory, and if successful append it
+     *  to a list of dirs to be deleted in the dtor. Since nestled dirs will
+     *  need to be deleted in reverse order to creation insert to the beginning
+     *  of the list.
```
Unnecessary implementation details here. Could we just say that "Creates a directory. If a directory is successfully created, it is removed when this object is destroyed"?

kursatkobya replied:
Those detailes are inherited from previous implementation, will change as you requested.

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,90 @@
+
+    /*@brief clear removes all elements from the m_createDirectoryCleanupHandlers
+     *
+     * */
```
Formatting

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,90 @@
+     * @brief tempDir Creates a uniquely named temporary directory according to templatePath.
+     *
+     * @warning The temporary path will be destroyed when the instance of FileToolkitWithUndo
+     *  is destroyed.
```
Which instance is this? When is it destroyed?

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,90 @@
+    bool createParentDirectory(const std::string path);
+
+    /**
+     * @brief m_createDirectoryCleanupHandlers holds cleanup handler for each created directory.
```
Is this comment about an old naming for the below member? I think it's enough to comment on the member and let it be implicit that the comment is about the member directly below the comment, to avoid mistakes like this.

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,83 @@
+     * @warning The created directory will be destroyed when the object is destroyed or rollBack
+     * method is called.
+     *
+     * @param templ a template Path used to create the path of the temporary directory, including
```
Typo in the param

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,83 @@
+     * @throws SoftwareContainerError if template path is not appropriate.
+     * @return A string path pointing to the newly created temporary directory.
+     */
+    std::string tempDir(std::string templatePath);
```
When does one use createDirectory and tempDir? Does tempDir create a directory?

kursatkobya replied:
tempDir and createDirectory are independent two functions. tempDir creates uniquely named temporary directory from template. 
Meanwhile createDirectory just creates a directory in given path. 
For more information one can look into bindMount function in FileToolkitWithUndo class.

breakreturn replied:
So "tempDir" should probably be named in a similar fashion as "createDirectory" is then? Otherwise the user will be forced to wonder why one says "create" and the other not, while both logically should create something.

And what do you mean with look into FileToolkitWithUndo? How would the user know that?

kursatkobya replied:
I didn't create those naming but i believe it is not the worst naming. Still i will change it to createTemplateDir which make more sense. 

These functions are not for users of SoftwareContainer but for developers of it. So if FileToolkitWithUndo does not ring any bells then probably the one doesn't need to have any  further look in this class!

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,83 @@
+    bool createParentDirectory(const std::string path);
+
+    /**
+     * @brief m_rollbackCleaners holds cleanup handler for each created directory.
```
Why mention the variable name here again? It makes it easy to miss when updating the code.

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -622,12 +625,24 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
```
You could probably rewrite this so that you use the same MountCleanupHandler as mentioned above (the one that cleans instantly), and then depending on result clean or save to `m_cleanupHandlers`.

kursatkobya replied:
I don't see any perk of doing like this. I prefer this way if it is not disruptive for you.

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @warning The created directory will be destroyed when the object is destroyed
+     *
+     * @param templatePath a template Path used to create the path of the temporary directory,
```
What's the template? What should one pass in here?

kursatkobya replied:
The last six characters of template must be XXXXXX and these are replaced with a string that makes the directory name unique. This one is mentioned within comment in paramline area "@param templatePath a template Path used to create the path of the temporary directory,
     * including "XXXXXX" which will be replaced with a unique ID for the temporary directory"

breakreturn commented on common/createdir.h:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @brief createParentDirectory Recursively tries to create the directory pointed to by path.
+     * @param path The directory path to be created.
+     * @return true on success, false on failure
```
Comment formatting is inconsistent with whitespace (compare this with the other comments).

tobsan commented on common/unit-test/createdir_unittest.cpp:
```diff
@@ -0,0 +1,81 @@
+        chmod("/tmp/test/error/in/nested", 444);
+        ASSERT_FALSE(tempCd.createDirectory("/tmp/test/error/in/nested/directory/error"));
+    }
+    ASSERT_FALSE(isDirectory("/tmp/test/error/in/nested"));
```
Imho this type of check should be in all the tests to make sure that things are rolled back. Or it can be part of the fixture class.
